Swarm docs: add administration guide for Managers and Raft

abronan opened the pull request:
Depends on #24207 and #24820 

**\- What I did**

Added some docs for `Manager` nodes administration and recommendations in _docker swarm mode_.

**\- How I did it**

Added some `.md` files under `/docs/swarm`.

**\- How to verify it**

By reading it and throwing some typo comments at me.

**\- Description for the changelog**

Added documentation on how to manage swarm Manager nodes

**\- A picture of a cute animal**

![bunny](https://cloud.githubusercontent.com/assets/1655859/16997623/15c6056e-4e6b-11e6-8e27-84e839b7f5fe.jpg)

Signed-off-by: Alexandre Beslic alexandre.beslic@gmail.com

/cc @sfsmithcha @mgoelzer

dperny commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,247 @@
+`docker node rm <id-node>` before adding a node back with a fresh state.
+
+In case of disaster recovery, you can take the data-directory of one of the
+members in order to restore the cluster with `--force-new-cluster
```
you dropped this:
```
`
```

abronan committed:
swarm docs: add administration guide for Managers and Raft

This adds a guide with recommendations on how to setup
Managers and keep the Swarm cluster healthy.

Signed-off-by: Alexandre Beslic <alexandre.beslic@gmail.com>

abronan commented:
@sfsmithcha pushed an update, I think it is OK as is, we can improve and iterate later on if there are details to add. Can you look into the docs failure (just `.md` links failures because it depends on other PRs). Thanks.

sfsmithcha commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,238 @@
+There is no limit to scale for manager nodes. The decision about how many
+manager nodes to implement is a trade-off between performance and
+fault-tolerance. Adding manager nodes to a swarm makes the swarm more
+fault-tolerant. However, additional manager nodes reduces write performance
```
typo s/reduces/reduce

sfsmithcha commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,238 @@
+
+An `unreachable` manager status on a Swarm manager means that this particular
+Manager is unreachable from other Managers. This means that action should be taken
+to restore this node to an healthy state (restarting the daemon, rebooting the machine...).
```
usage: "swarm" lowercase, "manager" lowercase. Suggest:

An manager with `unreachable` status means the manager node is unreachable from other manager nodes. In this case you need to take action to restore an odd number of managers:
* Restart the daemon and see if the manager comes back as reachable.
* Reboot the machine.
* If neither restarting or rebooting work, you should add another manager node or promote a worker to be a manager node.

sfsmithcha commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,238 @@
+<!--TODO what are the other states and what do they mean?-->
+
+From those commands, we can see that `node01` is both at the status
+`reachable` as a manager and `ready` as a worker.
```
Status and Manager status should match output. `Reachable`, `Ready`

sfsmithcha commented:
Docs CI error is expected until https://github.com/docker/docker/pull/24207 is merged.
Also, OK to publish with editorial comments. I can update in a later PR.
LGTM

abronan commented:
@sfsmithcha PTAL, tried to include your comments.

sfsmithcha commented:
LGTM

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+
+Swarm manager nodes use the [Raft Consensus Algorithm](raft.md) to manage the
+cluster state. You only need to understand some general concepts of the raft
+implementation in order to manage a swarm.
```
Nit: it's not really the implementation you need to understand

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+because more nodes must acknowledge proposals to update the cluster state.
+This means more network round-trip traffic.
+
+Raft requires a majority of members, also called a quorum of `(N/2)+1`, to
```
Insert a comma after quorum.

I'd remove `(N/2)+1`, or maybe state it in a separate sentence. Majority is clear, but "majority ... of (N/2)+1" might be confusing.

aaronlehmann replied:
Let's use the word "manager" instead of "member" throughout.

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+This means more network round-trip traffic.
+
+Raft requires a majority of members, also called a quorum of `(N/2)+1`, to
+agree on proposed updates to the cluster. A quorum must also agree on node
```
"A quorum of managers must also agree..."

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+
+You should maintain an odd number of managers in the swarm to support manager
+node failures. Having an odd number of nodes ensures that during a network
+partition, a quorum remains available to process requests.
```
Nitpicking again, but it takes more than odd `n` to ensure this:

- The network must be partitioned in two, not more than two, partitions
- All managers must be up
- n > 1

The point about an odd `n` being best is valid though. I would just rephrase this in a weaker way.

abronan replied:
Good point, thought about including that but kept the explanation simple for the first draft. I'll make the change.

abronan replied:
I took that into account in the first draft apparently..

```
a majority *could* stay available and be able to process our requests.
```

But I'll rephrase adding more details about network partitions though.

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+## Distribute manager nodes across datacenter regions
+
+In addition to maintaining an odd number of manager nodes, Docker recommends
+that you carefully place managers in the datacenter topology. For optimal
```
How about "In addition to maintaining an odd number of manager nodes, pay attention to datacenter topology when placing managers"?

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+fault-tolerance, distribute manager nodes across a minimum of 3 regions to
+support a region failure or maintenance scenario. If you suffer a failure in any
+region, the swarm still has a quorum of manager nodes available to process
+requests.
```
Really? Do you mean availability zones? I thought inter-region raft was problematic.

abronan replied:
Both work, it is only problematic if you encounter excessive latency between regions but you can tune the `ElectionTick` and `HeartbeatTick` to cope with the latency between regions you encounter issues. As long as Managers could contact each other through their IPs, it should not matter whether the communication is between availability zones or datacenter regions.

Maybe we should also mention availability-zones.

aaronlehmann replied:
If you don't have a majority in a single region, every raft write will take at least the round trip latency between the regions (times two, right? first to propagate the changes then mark them as committed). Since we serialize raft writes, this would be unusably slow for many applications.

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+
+By default manager nodes also act as a worker nodes. This means the scheduler
+can assign tasks to a manager node. For small and non-critical clusters
+assigning tasks to managers is relatively risk-free as long as you schedule
```
relatively low-risk

aaronlehmann commented:
Make sure to be consistent between the use of `raft` and `Raft` in these documents.

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+
+When you drain a node, the scheduler reassigns any tasks running on the node to
+other available worker nodes in the cluster. It also prevents the scheduler from
+assigning tasks on the node.
```
s/on/to/

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+recovery.
+
+You should never restart a manager node with the data directory from another
+node. For example, by copying the `raft` directory from one node to another. The
```
Put the example in parentheses or make it a complete sentence.

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+## Monitor swarm health
+
+You can monitor the health of Manager nodes by querying the docker `nodes` API
+in JSON format through the `/nodes` HTTP endpoint. Refer to the [nodes API documentation](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.24.md#36-nodes)
```
Is it proper to link to a document on github instead of the local version?

abronan replied:
/cc @sfsmithcha

sfsmithcha replied:
proper link here should be:  `../reference/api/docker_remote_api_v1.24.md#36-nodes`

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+for more information.
+
+From the command line, run `docker node inspect <id-node>` to query the nodes.
+For instance, to query the reachability of the node as a Manager to other raft cluster members:
```
manager should be lowercase. raft should probably be uppercase, but across all the documents.

aaronlehmann commented on docs/swarm/admin_guide.md:
````diff
@@ -0,0 +1,243 @@
+For instance, to query the reachability of the node as a Manager to other raft cluster members:
+
+```bash
+docker node inspect manager1 --format "{{ .ManagerStatus.Reachability }}"
````
Add a shell prompt

aaronlehmann commented on docs/swarm/admin_guide.md:
````diff
@@ -0,0 +1,243 @@
+To query the status of the node as a worker that accept tasks:
+
+```bash
+docker node inspect manager1 --format "{{ .Status.State }}"
````
Add a shell prompt

aaronlehmann commented on docs/swarm/admin_guide.md:
````diff
@@ -0,0 +1,243 @@
+docker node inspect manager1 --format "{{ .ManagerStatus.Reachability }}"
+reachable
+```
+<!--TODO what are the other states and what do they mean?-->
````
I would remove the `TODO` before merging. File an issue if necessary. A lot of people will read the markdown files in a text editor.

abronan replied:
@sfsmithcha added the TODO in an edit, I'm not sure what is the proper way to handle this in the docs. In doubt I'll remove it for now.

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+
+An `unreachable` health status means that this particular manager node is unreachable
+from other manager nodes. In this case you need to take action to restore an odd number of
+managers:
```
"...to restore the unreachable manager"?

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+Swarm is resilient to failures and the cluster can recover from any number
+of temporary node failures (machine reboots or crash with restart).
+
+In a swarm of `N` members, there must be a quorum of manager nodes greater than
```
`N` managers

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+failures.
+
+Even if you follow the guidelines here, it is possible that you can lose a
+quorum of manager nodes. If you can't recover the quorurm by conventional
```
typo: quorurm

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,243 @@
+
+The `--force-new-cluster` flag puts the Docker Engine into swarm mode as a
+manager node of a single-node cluster. It discards cluster membership information
+that existed before the loss of the quorum.
```
Mention that it retains data such as services, tasks, and the list of worker nodes.

aaronlehmann commented on docs/swarm/raft.md:
```diff
@@ -0,0 +1,47 @@
+When the Docker Engine runs in swarm mode, manager nodes implement the
+[Raft Consensus Algorithm](http://thesecretlivesofdata.com/raft/) to manage the global cluster state.
+
+The reason why *docker swarm mode* is using a Consensus algorithm is to make sure that
```
Uppercase D in Docker
Lowercase C in consensus

aaronlehmann commented on docs/swarm/raft.md:
```diff
@@ -0,0 +1,47 @@
+[Raft Consensus Algorithm](http://thesecretlivesofdata.com/raft/) to manage the global cluster state.
+
+The reason why *docker swarm mode* is using a Consensus algorithm is to make sure that
+all the Manager nodes that are in charge of managing and scheduling tasks in the cluster,
```
Lowercase Manager

aaronlehmann commented on docs/swarm/raft.md:
```diff
@@ -0,0 +1,47 @@
+the properties inherent to distributed systems:
+
+- *agreement on values* in a fault tolerant system. (Refer to [FLP impossibility theorem](http://the-paper-trail.org/blog/a-brief-tour-of-flp-impossibility/)
+ and the [Paxos Algorithm](http://www.cs.utexas.edu/users/lorenzo/corsi/cs380d/past/03F/notes/paxos-simple.pdf))
```
If we point to Paxos, let's make it clear we're not actually using Paxos, and this is just distributed systems education.

abronan commented:
Addressed the comments

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,242 @@
+cluster state. You only need to understand some general concepts of Raft in
+order to manage a swarm.
+
+There is no limit to scale for manager nodes. The decision about how many
```
Let's say "There is no limit on the number of manager nodes". The current phrasing makes it sound like adding more manager nodes is a way to scale the cluster.

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,242 @@
+because more nodes must acknowledge proposals to update the cluster state.
+This means more network round-trip traffic.
+
+Raft requires a majority of managers, also called a quorum to agree on proposed
```
Add a comma after quorum.

aaronlehmann commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,242 @@
+manager nodes across a minimum of 3 availability-zones or datacenter
+regions to support failures of an entire set of machines or common maintenance
+scenarios. If you suffer a failure in any of those zones, the swarm should maintain
+a quorum of manager nodes available to process requests and rebalance workloads.
```
This is a lot better but I still think we should put a warning that running Raft across regions may be extremely slow. Every write to Raft might take 200+ ms, and only one write can happen at a time.

aaronlehmann replied:
Also, since all control APIs and dispatcher APIs are proxied to the leader, the latency would be felt there too.

It would be interesting to actually test it with that kind of setup, and do store-intensive and dispatcher-intensive things like rolling updates.

abronan replied:
Added the warning, even though today's 2K nodes stress test was using managers scattered across DO regions (afaik) and it worked fine. The latency was a bit high on `docker service ls` after scheduling `96K` tasks but I think this was mostly due to checking the running replicas.

aluzzardi replied:
I'm with @aaronlehmann on this one. 

>  distribute manager nodes across a minimum of 3 availability-zones or datacenter regions

I would drop the "datacenter regions".

AZs were built specifically for this scenario. Even if it was feasible to run managers across regions, it's not practical - even if you have quorum, if you can't reach the other region, there's not much you can do. I'm assuming that no customer wants to have a EU container move to the US on failure.

aluzzardi commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,246 @@
+> infrastructure used to deploy the nodes. Write latencies could be high and issues
+> could arise with Leader election and Health Checks.
+
+| Swarm manager nodes |  Repartition (on 3 regions)  |
```
I would say "availability zones" rather than "regions"

aluzzardi commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,246 @@
+scenarios. If you suffer a failure in any of those zones, the swarm should maintain
+a quorum of manager nodes available to process requests and rebalance workloads.
+
+> Warning: Distributing managers between regions could be slow depending on the
```
Again, I'd get rid of the comment and just not talk about regions.

I think regions would be better served by federation rather than inter-region quorum

SvenDowideit commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,246 @@
+services using **resource constraints** for *cpu* and *memory*.
+
+However, because manager nodes use the Raft consensus algorithm to replicate data
+in a consistent way, they are sensitive to resouce starvation. You should
```
s/resouce/resource/ 

:yum: found one!! :)

SvenDowideit commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,246 @@
+The data directory is unique to a node ID and a node can only use a given node
+ID once to join the swarm. (ie. Node ID space should be globally unique)
+
+To cleanly re-join a manager node to a cluster:
```
this set of steps does not seem to apply to "backing up the cluster state" - does it deserve its own section?

SvenDowideit commented on docs/swarm/admin_guide.md:
```diff
@@ -0,0 +1,246 @@
+2. Run `docker node rm <id-node>` before adding a node back with a fresh state.
+3. Re-join the node to the cluster using `docker swarm join`.
+
+In case of [disaster recovery](#disaster-recovery), you can take the raft data
```
There's no heading of that name - and the section that does refer to disaster doesn't seem to mention how to use the backup data - I'd suggest putting "how to restore from backup" into this backup section.

thaJeztah commented:
@abronan can you squash your commits?

abronan commented:
@thaJeztah done

thaJeztah commented:
LGTM

thaJeztah commented:
thanks!
